File: impls/rust/src/step1_read_print.rs
mod reader;
mod types;

use std::io::Write;

use anyhow::Result;
use reader::read_str;
use types::MalVal;

fn read(input: &str) -> Result<MalVal> {
    read_str(input)
}

fn eval(ast: &MalVal) -> &MalVal {
    ast
}

fn print(val: &MalVal) -> String {
    val.pr_str(true)
}

fn rep(input: &str) -> String {
    match read(input) {
        Ok(ast) => print(eval(&ast)),
        Err(e) => e.to_string(),
    }
}

fn main() {
    let mut buffer = String::new();
    loop {
        print!("user> ");
        std::io::stdout().flush().unwrap();
        let input = match std::io::stdin().read_line(&mut buffer) {
            Ok(0) => {
                println!();
                break;
            }
            Err(_) => break,
            Ok(_) => buffer.trim(),
        };
        if !input.is_empty() {
            println!("{}", rep(input));
        }
        buffer.clear();
    }
}

File: impls/cpp/src/step1_read_print.cpp
#include <fmt/core.h>

#include <string>
#include <string_view>

#include "linenoise.h"
#include "printer.h"
#include "reader.h"
#include "types.h"

auto Read(std::string_view str) {
    try {
        return ReadStr(str);
    } catch (std::shared_ptr<MalType> err) {
        fmt::print("Exception {}", err->PrStr(false));
        return std::shared_ptr<MalType>(nullptr);
    }
}
auto Eval(auto ast) { return ast; }
std::string Print(auto ast) { return PrStr(ast, true); }
std::string Rep(std::string_view str) { return Print(Eval(Read(str))); }

int main() {
    linenoiseHistorySetMaxLen(30);
    linenoiseHistoryLoad("history.txt");

    char* line;
    while ((line = linenoise("user> ")) != NULL) {
        if (line[0] == '\0') continue;
        linenoiseHistoryAdd(line);
        try {
            fmt::print("{}\n", Rep(line));
        } catch (std::nullptr_t) {
            linenoiseFree(line);
            continue;
        }
        linenoiseFree(line);
    }
    linenoiseHistorySave("history.txt");
    return 0;
}
